File: SWEA/D3/2806. N－Queen/N－Queen.cpp
#include <iostream>

using namespace std;

int a[12];
int T,n,ret;

bool check(int col,int row) {
	for (int i = 0; i < row; i++) {
		if (col == a[i] || (row - i) == abs(col - a[i]))return false;
	}
	return true;
}

void nqueen(int row) {
	if (row == n) {
		ret++;
		return;
	}

	for (int i = 0; i < n; i++) {
		if (check(i, row)) {
			a[row] = i;
			nqueen(row + 1);
		}
	}
}

int main() {
	cin >> T;
	for (int tc = 1; tc <= T; tc++) {
		ret = 0;
		cin >> n;
		nqueen(0);

		cout << "#" << tc << " " << ret << '\n';
	}
}

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.util.Scanner;
import java.lang.*;

import static java.lang.Math.abs;

public class Main{
    static int n,ret;

    static int[] arr = new int[15];

    static boolean check(int row,int col){
        for (int i=0; i<col; i++){
            if(arr[i] == row)return false;
            if ((col-i) == abs(row-arr[i]))return false;
        }
        return true;
    }

    static void nqueen(int col){
        if (col == n){
            ret++;
            return;
        }

        for (int i=0; i<n;i++) {
            if (check(i,col)){
                arr[col] = i;
                nqueen(col+1);
            }
        }
    }

    public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        nqueen(0);
        System.out.println(ret);
    }
}